src/couch_views/src/couch_views_reader.erl

```diff
@@ -0,0 +1,204 @@
+            % along with the docs we would skip.
+            % Limit = (Doc limit + Skip) * Num of Rows per Map KV
+            [{limit, (Limit + Skip) * 2} | Opts]
+    end,
+
+    Opts2 = case Direction of
+        fwd ->
+            Opts1;
+        rev ->
+            [{reverse, true} | Opts1]
+    end,
+    {Start1, End1, Opts2}.
+
+
+get_range_keys(Db, Sig, IdxName, Args) ->
+    #{
+        inclusive_end := InclusiveEnd,
+        direction := Direction
+    } = Args,
+
+    {MapStartKey, MapEndKey} = case Direction of
+        fwd -> {start_key, end_key};
+        rev -> {end_key, start_key}
+    end,
+
+    {Start0, End0} = couch_views_fdb:get_map_range_keys(Db, Sig, IdxName),
+
+    Start1 = case maps:is_key(MapStartKey, Args) of
+        false ->
+            Start0;
+        true ->
+            StartKey = maps:get(MapStartKey, Args),
+            Start = couch_views_fdb:get_map_index_key(Db, Sig, IdxName,
+                StartKey),
+            erlfdb_key:first_greater_or_equal(Start)
+    end,
+
+    End1 = case maps:is_key(MapEndKey, Args) of
+        false ->
+            End0;
+        true ->
+            EndKey = maps:get(MapEndKey, Args),
+            EndBin = couch_views_fdb:get_map_index_key(Db, Sig, IdxName,
+                EndKey),
+            EndBin1 = case InclusiveEnd of
+                true -> <<EndBin/binary, 16#FF>>;
+                false -> EndBin
+            end,
+            erlfdb_key:first_greater_than(EndBin1)
+    end,
+    {Start1, End1}.
+
+
+get_unpack_fun(TxDb, Opts, Callback) ->
+    UnPackFwd = fun({K, V}, State) ->
+        case couch_views_fdb:unpack_map_row(TxDb, K, V) of
+            {key, _Id, RowKey} ->
+                maps:put(current_key, RowKey, State);
+            {value, Id, RowValue} ->
+                #{
+                    current_key := RowKey,
+                    acc := Acc,
+                    skip := Skip,
+                    db := Db
+                } = State,
+
+                case Skip > 0 of
+                    true ->
+                        maps:put(skip, Skip - 1, State);
+                    false ->
+                        Row = [{id, Id}, {key, RowKey}, {value, RowValue}],
+
+                        IncludeDoc = maps:get(include_docs, State, false),
+                        Row1 = maybe_include_doc(Db, Id, Row, IncludeDoc),
+
+                        {ok, AccNext} = Callback({row, Row1}, Acc),
+                        maps:put(acc, AccNext, State)
+                end
+        end
+    end,
+
+    UnPackRev = fun({K, V}, State) ->
+        case couch_views_fdb:unpack_map_row(TxDb, K, V) of
+            {key, Id, RowKey} ->
+                #{
+                    current_value := RowValue,
+                    acc := Acc,
+                    skip := Skip,
+                    db := Db
+                } = State,
+
+                case Skip > 0 of
+                    true ->
+                        maps:put(skip, Skip - 1, State);
```

Comment:
Minor nit. I usually use `State#{skip := Skip -1}` syntax when updating, and we can also assert that the value already exists (with `:=`). But up to you, if you prefer, `maps:put`, that's ok too

Reply:
Awesome. I didn't know you could do that with maps. I've changed them. It makes more sense.